File: week_11/Red_Green.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/imgproc.hpp>
#include <iostream>
#include <vector>

using namespace std;

// 定义颜色范围
struct ColorRange {
    cv::Scalar lower;
    cv::Scalar upper;
};

// 检测颜色并返回中心点
cv::Point2f detectColor(const cv::Mat& hsv, const ColorRange& color_range) {
    cv::Mat mask;
    cv::inRange(hsv, color_range.lower, color_range.upper, mask);

    vector<vector<cv::Point>> contours;
    cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    cv::Point2f center(0, 0);
    if (!contours.empty()) {
        int max_area_idx = 0;
        double max_area = 0;
        for (size_t i = 0; i < contours.size(); ++i) {
            double area = cv::contourArea(contours[i]);
            if (area > max_area) {
                max_area = area;
                max_area_idx = i;
            }
        }

        if (max_area_idx >= 0) {
            const auto& contour = contours[max_area_idx];
            cv::Rect rect = cv::boundingRect(contour);
            center.x = rect.x + rect.width / 2.0;
            center.y = rect.y + rect.height / 2.0;
        }
    }

    return center;
}

// 绘制点
void drawPoints(cv::Mat& image, const cv::Point2f& center, const std::string& label) {
    cv::putText(image, label, center, cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 255), 2);
}

int main() {
    cv::Mat image = cv::imread("/home/nhmgs/cv2/week_11/src/img_3.jpeg");
    if (image.empty()) {
        cout << "未找到该图片" << endl;
        return -1;
    }

    cv::Mat hsv;
    cv::cvtColor(image, hsv, cv::COLOR_BGR2HSV);

    ColorRange green_range = {{36, 100, 100}, {86, 255, 255}};
    cv::Point2f green_center = detectColor(hsv, green_range);
    drawPoints(image, green_center, "Green point");

    ColorRange red_range = {{165, 100, 100}, {179, 255, 255}};
    cv::Point2f red_center = detectColor(hsv, red_range);
    drawPoints(image, red_center, "Red point");

    cv::namedWindow("image", cv::WINDOW_NORMAL);
    cv::imshow("image", image);
    // cv::waitKey(0);
    // cv::destroyAllWindows();

    return 0;
}

File: week_9/Final_Code/3.0 实时显示/Red_Green.py
import cv2
import numpy as np

def detect_color(hsv, lower_color, upper_color):

    mask_color = cv2.inRange(hsv, lower_color, upper_color)
    contours_color, _ = cv2.findContours(mask_color, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    if len(contours_color) > 0:
        max_contour = max(contours_color, key=cv2.contourArea)
        x, y, w, h = cv2.boundingRect(max_contour)
        center_x = x + w // 2
        center_y = y + h // 2
    else:
        center_x, center_y = 0, 0
    print('center_x:', center_x, 'center_y:', center_y)

    return center_x, center_y

def detect_green(image):

    hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
    lower_green = np.array([36, 100, 100])
    upper_green = np.array([86, 255, 255])
    
    return detect_color(hsv, lower_green, upper_green)

def detect_red(image):
    
    hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
    lower_red = np.array([165, 100, 100])
    upper_red = np.array([179, 255, 255])
    
    return detect_color(hsv, lower_red, upper_red)
def draw_points(image, center_x, center_y):
  
    cv2.putText(image, f'point: ({center_x}, {center_y})', (center_x, center_y - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 2)

def find_red_green(image):
    center_x, center_y = detect_red(image)
    draw_points(image, center_x, center_y)
    center_x, center_y = detect_green(image)
    draw_points(image, center_x, center_y)
    
    return image

if __name__ == '__main__': 

    image = cv2.imread('week_9/src/img_2.jpeg')
    image = find_red_green(image)
    cv2.namedWindow('image',cv2.WINDOW_NORMAL)
    cv2.imshow('image', image)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
